src/record/partially_serialized.rs

```diff
@@ -3,20 +3,56 @@ use crate::prelude::*;
 use bytes::{Bytes, BytesMut};
 
 pub(crate) struct PartiallySerializedWriteResult {
-    bytes_written: u64,
+    blob_offset: u64,
     header_checksum: u32,
 }
 
 impl PartiallySerializedWriteResult {
-    pub(crate) fn bytes_written(&self) -> u64 {
-        self.bytes_written
+    pub(crate) fn blob_offset(&self) -> u64 {
+        self.blob_offset
     }
 
     pub(crate) fn header_checksum(&self) -> u32 {
         self.header_checksum
     }
 }
 
+struct RecordWritableDataCreator {
+    record: PartiallySerializedRecord,
+    len: u64,
+}
+
+impl From<PartiallySerializedRecord> for RecordWritableDataCreator {
+    fn from(record: PartiallySerializedRecord) -> Self {
+        let len = record.head_with_data.len() as u64
+            + record.data.as_ref().map(|v| v.len()).unwrap_or(0) as u64;
+        Self { record, len }
+    }
+}
+
+impl WritableDataCreator<(u64, u32)> for RecordWritableDataCreator {
```

Comment:
Looks like you can implement `WritableDataCreator<(u64, u32)>` directly on `PartiallySerializedRecord`

Reply:
Additionally: you already have `PartiallySerializedWriteResult` that contains `offset` + `checksum`, so you can use that type instead of tuple